sympy/sets/powerset.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import print_function, division
+
+from sympy.core.decorators import _sympifyit
+from sympy.core.evaluate import global_evaluate
+from sympy.core.logic import fuzzy_bool
+from sympy.core.singleton import S
+from sympy.core.sympify import _sympify
+
+from .sets import Set, tfn
+
+
+class PowerSet(Set):
+    r"""A symbolic object representing a power set.
+
+    Parameters
+    ==========
+
+    arg : Set
+        The set to take power of.
+
+    evaluate : bool
+        The flag to control evaluation.
+
+        If the evaluation is disabled for finite sets, it can take
+        advantage of using subset test as a membership test.
+
+    Notes
+    =====
+
+    Power set `\mathcal{P}(S)` is defined as a set containing all the
+    subsets of `S`.
+
+    If the set `S` is a finite set, its power set would have
+    `2^{\left| S \right|}` elements, where `\left| S \right|` denotes
+    the cardinality of `S`.
+
+    Examples
+    ========
+
+    >>> from sympy.sets.powerset import PowerSet
+    >>> from sympy import S, FiniteSet
+
+    A power set of a finite set:
+
+    >>> PowerSet(FiniteSet(1, 2, 3))
+    {EmptySet(), {1}, {2}, {3}, {1, 2}, {1, 3}, {2, 3}, {1, 2, 3}}
+
+    A power set of an empty set:
+
+    >>> PowerSet(S.EmptySet)
+    {EmptySet()}
+    >>> PowerSet(PowerSet(S.EmptySet))
+    {EmptySet(), {EmptySet()}}
+
+    A power set of an infinite set:
+
+    >>> PowerSet(S.Reals)
+    PowerSet(Reals)
+
+    An unevaluated power set:
+
+    >>> PowerSet(FiniteSet(1, 2, 3), evaluate=False)
+    PowerSet({1, 2, 3})
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Power_set
+
+    .. [2] https://en.wikipedia.org/wiki/Axiom_of_power_set
+    """
+    def __new__(cls, arg, evaluate=global_evaluate[0]):
+        arg = _sympify(arg)
+
+        if not isinstance(arg, Set):
+            raise ValueError('{} must be a set.'.format(arg))
+
+        if evaluate:
+            ret = arg._eval_powerset()
+
+            if ret is not None and not isinstance(arg, PowerSet):
+                return ret
+
+        return super(PowerSet, cls).__new__(cls, arg)
+
+    @property
+    def arg(self):
+        return self.args[0]
+
+    @_sympifyit('other', NotImplemented)
+    def _contains(self, other):
+        if not isinstance(other, Set):
+            raise ValueError('{} must be a set.'.format(other))
```

Comment:
This line isn't tested.

Reply:
This isn't consistent with other sets:
```julia
In [7]: PowerSet({1})._contains(2)                                                                                                

In [8]: PowerSet({1}, evaluate=False)._contains(2)                                                                                
---------------------------------------------------------------------------
ValueError
```

Reply:
`_contains(2)` should evaluate to False. Looks like EmptySet needs `_eval_Eq`:
```julia
In [7]: P = PowerSet({1})                                                                                                         

In [8]: P                                                                                                                         
Out[8]: {∅, {1}}

In [9]: print(P._contains(2))                                                                                                     
None

In [10]: Eq(S.EmptySet, 2)                                                                                                        
Out[10]: ∅ = 2
```

Reply:
I don't think that the relations like below are well defined across various set theories. An element who cannot be a set would be called urelement, but it is not shared across every axioms.

![image](https://user-images.githubusercontent.com/34944973/64411759-e5443580-d0c8-11e9-88fb-f7571c0d191c.png)

![image](https://user-images.githubusercontent.com/34944973/64411794-fc832300-d0c8-11e9-8b9f-db196f8daf68.png)

![image](https://user-images.githubusercontent.com/34944973/64411822-04db5e00-d0c9-11e9-8ed0-8c4462c2a3e8.png)

But only these statements may be determined to be false

![image](https://user-images.githubusercontent.com/34944973/64411839-102e8980-d0c9-11e9-8e35-c04fb89f280c.png)

![image](https://user-images.githubusercontent.com/34944973/64411879-263c4a00-d0c9-11e9-8fc8-57b67de71c88.png)